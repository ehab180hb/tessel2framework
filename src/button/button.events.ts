import { greenButton, redButton, blueButton } from '../button/button.handler'
import { BoardEvents, boardEvents } from '../board/board.events'

interface UpdateButtonsInput {
  greenButtonCb?: () => void
  redButtonCb?: () => void
  blueButtonCb?: () => void
}

export const attachButtonUpdate = (updateButtonsParams: UpdateButtonsInput): void => {
  const emptyButtonCb = () => {}
  const {
    greenButtonCb = emptyButtonCb,
    redButtonCb = emptyButtonCb,
    blueButtonCb = emptyButtonCb,
  } = updateButtonsParams
  greenButton(greenButtonCb)
  redButton(redButtonCb)
  blueButton(blueButtonCb)
}

export const updateButtons = (updateButtonsParams: UpdateButtonsInput) => {
  boardEvents.emit(BoardEvents.UPDATE_BUTTONS, updateButtonsParams)
}
