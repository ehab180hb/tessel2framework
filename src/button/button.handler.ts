import * as five from 'johnny-five'
import { PIN } from '../board/board.model'

const getButton = (pin: PIN) => (cb: () => void) => new five.Button(pin).on('press', cb)

export const redButton = getButton(PIN.b5)
export const greenButton = getButton(PIN.b6)
export const blueButton = getButton(PIN.b7)
