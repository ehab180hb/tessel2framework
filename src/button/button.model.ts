import * as five from 'johnny-five'
import { PIN } from '../board/board.model'

export interface ButtonCall {
  pin: PIN
  name: string
}

export interface DeviceButton {
  name: string
  pin: PIN
  listener: five.Button
}
