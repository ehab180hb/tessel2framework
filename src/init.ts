import { boardEvents, BoardEvents } from './board/board.events'
import { attachScreenUpdate } from './lcd/lcd.events'
import { attachButtonUpdate } from './button/button.events'

export const initComponents = () => {
  boardEvents.on(BoardEvents.UPDATE_SCREEN, attachScreenUpdate)
  boardEvents.on(BoardEvents.UPDATE_BUTTONS, attachButtonUpdate)
}
