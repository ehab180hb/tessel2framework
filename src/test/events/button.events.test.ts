import * as five from 'johnny-five'
import { attachButtonUpdate, updateButtons } from '../../button/button.events'
import { initComponents } from '../../init'

describe('Button events', () => {
  let buttonSpy: string[] = []
  const mockButtonAction = () => {}
  const mockButtonAction2 = () => {}

  describe('attachButtonUpdate()', () => {
    beforeEach(() => {
      buttonSpy = []
      ;(five.Button as any) = jest.fn(pin => ({
        on: (e: any, fn: Function) => {
          buttonSpy.push(JSON.stringify({ e, fn: fn.name, pin }))
        },
      }))
    })
    test('Works with default parameters', () => {
      attachButtonUpdate({})
      expect(buttonSpy).toMatchSnapshot()
    })
    test('Works with custom parameters', () => {
      attachButtonUpdate({
        greenButtonCb: mockButtonAction,
        blueButtonCb: mockButtonAction,
        redButtonCb: mockButtonAction,
      })
      expect(buttonSpy).toMatchSnapshot()
    })
  })

  describe('updateButtons()', () => {
    test('Works with custom parameters', () => {
      initComponents()
      updateButtons({
        greenButtonCb: mockButtonAction2,
        blueButtonCb: mockButtonAction2,
        redButtonCb: mockButtonAction2,
      })
      expect(buttonSpy).toMatchSnapshot()
    })
  })
})
