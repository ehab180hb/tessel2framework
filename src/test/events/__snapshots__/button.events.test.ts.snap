// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button events attachButtonUpdate() Works with custom parameters 1`] = `
Array [
  "{\\"e\\":\\"press\\",\\"fn\\":\\"mockButtonAction\\",\\"pin\\":14}",
  "{\\"e\\":\\"press\\",\\"fn\\":\\"mockButtonAction\\",\\"pin\\":13}",
  "{\\"e\\":\\"press\\",\\"fn\\":\\"mockButtonAction\\",\\"pin\\":15}",
]
`;

exports[`Button events attachButtonUpdate() Works with default parameters 1`] = `
Array [
  "{\\"e\\":\\"press\\",\\"fn\\":\\"emptyButtonCb\\",\\"pin\\":14}",
  "{\\"e\\":\\"press\\",\\"fn\\":\\"emptyButtonCb\\",\\"pin\\":13}",
  "{\\"e\\":\\"press\\",\\"fn\\":\\"emptyButtonCb\\",\\"pin\\":15}",
]
`;

exports[`Button events updateButtons() Works with custom parameters 1`] = `
Array [
  "{\\"e\\":\\"press\\",\\"fn\\":\\"mockButtonAction\\",\\"pin\\":14}",
  "{\\"e\\":\\"press\\",\\"fn\\":\\"mockButtonAction\\",\\"pin\\":13}",
  "{\\"e\\":\\"press\\",\\"fn\\":\\"mockButtonAction\\",\\"pin\\":15}",
  "{\\"e\\":\\"press\\",\\"fn\\":\\"mockButtonAction2\\",\\"pin\\":14}",
  "{\\"e\\":\\"press\\",\\"fn\\":\\"mockButtonAction2\\",\\"pin\\":13}",
  "{\\"e\\":\\"press\\",\\"fn\\":\\"mockButtonAction2\\",\\"pin\\":15}",
]
`;
